File: 2022/day07/src/main.rs
use std::collections::HashMap;

#[derive(Debug)]
enum DirEnt {
    File(usize),
    Directory(HashMap<String, DirEnt>),
}

impl DirEnt {
    fn add_entry(&mut self, spec: &str) {
        if let DirEnt::Directory(ref mut map) = self {
            let parts: Vec<&str> = spec.split(' ').collect();
            assert!(parts.len() == 2);
            let name = String::from(parts[1]);
            map.insert(
                name,
                match parts[0] {
                    "dir" => DirEnt::Directory(HashMap::new()),
                    _ => DirEnt::File(parts[0].parse().unwrap()),
                },
            );
        } else {
            panic!("DirEnt is not a directory");
        }
    }

    fn cd(&mut self, name: &str) -> &mut DirEnt {
        if let DirEnt::Directory(ref mut map) = self {
            return map.get_mut(name).unwrap();
        } else {
            panic!("DirEnt is not a directory");
        }
    }
}

// https://blog.rust-lang.org/2015/04/17/Enums-match-mutation-and-moves.html

fn build(input: &String) -> DirEnt {
    let mut root = DirEnt::Directory(HashMap::new());
    let mut path: Vec<&mut DirEnt> = vec![&mut root];
    for line in input.lines() {
        if line.as_bytes()[0] == '$' as u8 {
            let parts: Vec<&str> = line.split(' ').collect();
            if parts.len() == 3 {
                assert!(parts[0] == "$");
                assert!(parts[1] == "cd");
                if parts[2] == ".." {
                    path.pop();
                } else if parts[2] != "/" {
                    let lasti = path.len() - 1;
                    let cwd = &mut path[lasti];
                    path.push(cwd.cd(parts[2]));
                }
            }
        } else {
            let lasti = path.len() - 1;
            let cwd = &mut path[lasti];
            println!("Adding: {} to {:?}", line, cwd);
            cwd.add_entry(line);
        }
    }
    root
}

fn first(input: &String) {
    let root = build(input);
}

fn second(input: &String) {}

fn main() {
    let input = std::fs::read_to_string("input.txt").unwrap();
    first(&input);
    second(&input);
}

File: 2022/day07/src/main.py
#!/usr/bin/env python3

class DirEnt:
  def __init__(self, size=-1, isdir=False) -> "DirEnt":
    self.isdir = isdir
    self.size = size
    if isdir:
      self.children = {}
    else:
      self.children = None

  def __repr__(self):
    return "DirEnt(" + (f"children={self.children}" if self.isdir else f"size={self.size}") + ")"

def build(input_path: str) -> DirEnt:
  root = DirEnt(isdir=True)
  path = [root]
  numlines = 0
  for line in open(input_path, "r"):
    line = line.strip()
    numlines += 1
    if numlines == 100:
      print(f"{path}, {len(path)}")
    parts = line.split(" ")
    if parts[0] == "$":
      if parts[1] == "cd":
        if parts[2] == "..":
          path.pop()
        elif parts[2] != "/":
          cwd = path[-1]
          path.append(cwd.children[parts[2]])
      else:
        assert parts[1] == "ls", f"{parts}"
        assert len(parts) == 2, f"{parts}"
    else:
      assert len(parts) == 2, f"{parts}"
      cwd = path[-1]
      if parts[0] == "dir":
        isdir = True
        size = -1
      else:
        isdir = False
        size = int(parts[0])
      cwd.children[parts[1]] = DirEnt(size=size, isdir=isdir)
  return root

def find_sizes(ent: DirEnt, sizes: list, depth: int = 0) -> int:
  if ent.isdir:
    sum = 0
    for name, child in ent.children.items():
      size = find_sizes(child, sizes, depth + 1)
      print(("  " * depth) + f"{name}: {size}")
      sum += size
    print(("  " * depth) + f"= {sum}")
    sizes.append(sum)
    return sum
  else:
    return ent.size

def main(input_path: str):
  root = build(input_path)
  sizes = []
  find_sizes(root, sizes)
  sum = 0
  for size in sizes:
    if size <= 100000:
      print(f"+ {size}")
      sum += size
  print(f"sum <= 10000: {sum}")

if __name__ == "__main__":
  # main(input_path="input.txt")
  main(input_path="input.txt")
